clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```
@@ -26,35 +28,165 @@
+        private final int maxSize;
+        private final ByteBuffer startBuffer;
+        private int dataSize = -1;
+        private short apiKey = (short) -1;
+        private short apiVersion = (short) -1;
+
+        public RequestStartHeaderReceiver(int maxSize) {
+            this.maxSize = maxSize;
+            this.startBuffer = ByteBuffer.allocate(8);
+        }
+
+        @Override
+        public ByteBuffer allocatePayload() {
+            if (dataSize < 0)
+                return null;
+            ByteBuffer buffer = ByteBuffer.allocate(dataSize);
+            buffer.putShort(apiKey);
+            buffer.putShort(apiVersion);
+            return buffer;
+        }
 
-    public NetworkReceive(String source, ByteBuffer buffer) {
+        @Override
+        public int read(ReadableByteChannel channel) throws IOException {
+            if (dataSize > 0)
+                return 0;
+            int read = channel.read(startBuffer);
+            if (read < 0)
+                throw new EOFException();
+            if (startBuffer.hasRemaining())
+                return read;
+            int size = startBuffer.getInt();
+            if (size < MIN_REQUEST_SIZE) {
+                throw new InvalidReceiveException("Invalid request receive (size = " + size + ")");
+            } else if (size > maxSize) {
+                throw new InvalidReceiveException("Invalid request receive (size = " + size + " larger than " + maxSize + ")");
+            }
+            apiKey = startBuffer.getShort();
+            ApiKeys api;
+            apiVersion = startBuffer.getShort();
+            try {
+                api = ApiKeys.forId(apiKey);
+            } catch (IllegalArgumentException e) {
+                throw new InvalidReceiveException("Unknown API key " + apiKey);
+            }
+            if (apiVersion < 0) {
+                throw new InvalidReceiveException("Invalid api version " + apiVersion);
+            } else if (api.latestVersion() < apiVersion) {
```

Should we also throw InvalidReceiveException if apiVersion is less than oldestVersion?